MAINT Introduce Pairwise Distances Reductions private submodule

jjerphan opened the pull request:
#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Part of splitting https://github.com/scikit-learn/scikit-learn/pull/21462#issuecomment-993862788

⚠ This targets the `upstream:pairwise-distances-argkmin` feature branch not `upstream:main`.

#### What does this implement/fix? Explain your changes.

This introduces the necessary private implementations for a new
private submodule, i.e.:

 - DatasetsPair, an abstraction to wrap a pair of two datasets and
 compute their vectors pairwise distances
 - DenseDenseDatasetsPair, a first implementation of DatasetsPair
 for pair of two dense datasets
 - PairwiseDistancesReduction, an abstraction allowing computing
 reductions efficiently in parallel and of
 - PairwiseDistancesArgkmin, a first implementation of
 PairwiseDistancesReduction for k-Nearest Neighbors search


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jjerphan committed:
MAINT Introduce Pairwise Distances Reductions private submodule

This introduces the neccessary private implementations for a new
private submodule, i.e.:

 - DatasetsPair, an abstraction to wrap a pair of two datasets and
 compute their vectors pairwise distances
 - DenseDenseDatasetsPair, a first implementation of DatasetsPair
 for pair of two dense datasets
 - PairwiseDistancesReduction, an abstraction allowing computing
 reductions efficiently in parallel and of
 - PairwiseDistancesArgkmin, a first implementation of
 PairwiseDistancesReduction for k-Nearest Neighbors search

jjerphan committed:
Retrigger CI for failing Circle CI job

jjerphan committed:
TST Improve _get_dummy_metric_params_list

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

lorentzenchr commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,158 @@ cdef class PyFuncDistance(DistanceMetric):
+    aggregation logic from metric-specific computation as much as
+    possible.
+
+    X and Y can be stored as np.ndarrays or CSR matrices in subclasses.
```
```suggestion
    X and Y can be stored as C-contiguous np.ndarrays or CSR matrices in subclasses.
```

lorentzenchr commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,158 @@ cdef class PyFuncDistance(DistanceMetric):
+                **(metric_kwargs or {})
+            )
+
+        if X.dtype != np.float64 or Y.dtype != np.float64:
```
```suggestion
        if not (X.dtype == Y.dtype == np.float64):
```
very optional suggestion

jjerphan replied:
I find this way more elegant.

lorentzenchr commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,158 @@ cdef class PyFuncDistance(DistanceMetric):
+
+        return DenseDenseDatasetsPair(X, Y, distance_metric)
+
+    @classmethod
+    def unpack_csr_matrix(cls, X: csr_matrix):
+        """Ensure getting ITYPE instead of int internally used for CSR matrices."""
+        X_data = np.asarray(X.data, dtype=DTYPE)
+        X_indices = np.asarray(X.indices, dtype=ITYPE)
+        X_indptr = np.asarray(X.indptr, dtype=ITYPE)
+        return X_data, X_indptr, X_indptr
+
```
```suggestion
```
Add this when we add `SparseSpraseDatasetsPair`.

lorentzenchr commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,158 @@ cdef class PyFuncDistance(DistanceMetric):
+    def __init__(self, DistanceMetric distance_metric):
+        self.distance_metric = distance_metric
+
+    cdef ITYPE_t n_samples_X(self) nogil:
+        """Number of samples in X."""
+        return -999
```
Is it possible to place the implementation here? It should work for dense and sparse. I just don't know about cython.final.
And the same for `n_samples_Y`.
```suggestion
    @final
    cdef ITYPE_t n_samples_X(self) nogil:
        return self.X.shape[0]
```
Otherwise, a comment stating "This is like an abstract base class method and should never be used. TODO: use with gil: raise in Cython 3.0"

jjerphan replied:
It's not possible because `X` and `Y` aren't attribute for `DatasetsPair` (though they do for `DenseDenseDatasetsPair`).

I'll thus add the comment you suggest.

lorentzenchr commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,158 @@ cdef class PyFuncDistance(DistanceMetric):
+
+@final
+cdef class DenseDenseDatasetsPair(DatasetsPair):
+    """Compute distances between vectors of two arrays.
```
```suggestion
    """Compute distances between row vectors of two arrays.
```

lorentzenchr commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,158 @@ cdef class PyFuncDistance(DistanceMetric):
+
+    distance_metric: DistanceMetric
+        The distance metric responsible for computing distances
+        between two vectors of (X, Y).
```
```suggestion
        between two row vectors of (X, Y).
```

lorentzenchr reviewed (commented):
A review round about _dist_metrics.pyx.

lorentzenchr commented on sklearn/metrics/setup.py:
```diff
@@ -26,6 +26,12 @@ def configuration(parent_package="", top_path=None):
         libraries=libraries,
     )
 
+    config.add_extension(
+        "_pairwise_distances_reduction",
+        sources=["_pairwise_distances_reduction.pyx"],
+        libraries=libraries,
+    )
```
As we `cimport numpy`, I'm surprised that we do not need the usual `include_dirs=[np.get_include(), os.path.join(np.get_include(), "numpy")]`.

jjerphan replied:
I think this unexpectedly works because other extensions defined in this setup for `metrics` do specify numpy's include folders (IIRC/IIUC, internally Cython performs a first pass on all `pxd` headers prior to compiling sources).

For safety because `_pairwise_distances_reduction` needs the numpy's include folders, I will specify `include_dirs` here.

lorentzenchr reviewed (commented):
Review on setup:wink:

jjerphan committed:
Address review comments

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+            "hamming",
+            *BOOL_METRICS,
+        }
+        return sorted(set(METRIC_MAPPING.keys()).difference(excluded))
```
```suggestion
        return sorted(set(METRIC_MAPPING.keys()) - excluded)
```

jjerphan replied:
Today I learned something.

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+    datasets_pair: DatasetsPair
+        The pair of dataset to use.
+
+    chunk_size: int, default=None,
```
```suggestion
    chunk_size: int, default=None
```

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+    """Abstract base class for pairwise distance computation & reduction
+
+    Subclasses of this class compute pairwise distances between a set of
+    vectors (rows) X and another set of vectors (rows) Y and apply a
```
```suggestion
    row vectors of X and another set of row vectors of Y and apply a
```
And several times below. I think "row vector" is a good term.

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+        True if the PairwiseDistancesReduction can be used, else False.
+        """
+        # Coercing to np.array to get the dtype
+        # TODO: what is the best way to get lists' dtype?
```
Is it necessary to do the check here? Or can we delegate it to the caller of this and then expect a `check_array` to have already happend?
Or can we call `check_array` here?

jjerphan replied:
This is to be address in a following sub-PR.

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+          `pairwise_dist_parallel_strategy`, and use 'auto' if it is not set.
+    """
+
+    cdef:
+        readonly DatasetsPair datasets_pair
```
Why are all these defined as class/static variables instead of instance variables (set in __init__)?

jjerphan replied:
For the sake of clarity (because I do not know what you are referring to using "these"), I will be verbose:

All the definition under `cdef`-scope are [instance attributes definitions](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#introduction) not class attributes definitions.

Moreover,
```
cdef readonly DatasetsPair datasets_pair
```
is used to have [this attribute accessible from Python scopes](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#static-attributes).

Does that answer your question?

lorentzenchr replied:
Yes. Also, I was confused. It's fine as it is.

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+            ITYPE_t n_samples_chunk, X_n_full_chunks, Y_n_full_chunks
+
+        if chunk_size is None:
+            chunk_size = get_config().get("pairwise_dist_chunk_size", 256)
```
If this is the only place where `256` enters the implementation (not counting docstrings and comments) this is fine. Otherwise I would prefer a compile time constant `DEF ...`.

jjerphan replied:
Other reviewers do not agree on this.
Personally I do not have any opinion.

I would prefer sticking to this (not to go into circles again).

lorentzenchr replied:
This seems to be the only place, for know. So we are good. For the future, ping me if you need support to argue in favor of a compile time constant.

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+
+            for X_chunk_idx in prange(self.X_n_chunks, schedule='static'):
+                X_start = X_chunk_idx * self.X_n_samples_chunk
+                if (X_chunk_idx == self.X_n_chunks - 1
+                    and self.X_n_samples_remainder > 0):
+                    X_end = X_start + self.X_n_samples_remainder
+                else:
+                    X_end = X_start + self.X_n_samples_chunk
```
This is a pattern that occurs 4 times in the code. Can you make an inline cdef helper function for it?

jjerphan replied:
I think it's easier keeping this as is: the logic is simple enough and introducing a helper method for this would be uneasy as two values would need to be returned.

lorentzenchr replied:
In order to return a pair of values, have a look at https://github.com/scikit-learn/scikit-learn/blob/5bb18166d420c5e265844c32442d960ca0c5005e/sklearn/_loss/_loss.pxd#L21-L24 and then a function like `cy_grad_hess`.

I thought, it would save a little code and make reading it easier. No hard feelings, if you won't pursue this suggestion.

jjerphan replied:
Let's wait for someone else's opinion.

thomasjpfan replied:
I think we can do a slight refactor of `__init__` to not need a helper here:

```python
X_n_samples_remainder = self.n_samples_X % self.X_n_samples_chunk
self.X_n_chunks = X_n_full_chunks + (X_n_samples_remainder != 0)

if X_n_samples_remainder != 0:
    self.X_n_samples_last_chunk = X_n_samples_remainder
else:
    self.X_n_samples_last_chunk = X_n_samples_chunk
```

In the for loop we can now have:

```python
for X_chunk_idx in prange(self.X_n_chunks, schedule='static'):
    X_start = X_chunk_idx * self.X_n_samples_chunk
    if X_chunk_idx == self.X_n_chunks - 1:
        X_end = X_start + self.X_n_samples_last_chunk
    else:
        X_end = X_start + self.X_n_samples_chunk
```

I think only checking for the last index is simple enough to keep inline.

jjerphan replied:
Addressed in a2f7b6d11597388e95443325d692037ef6eb321c.

Please, let me know what you think of it.

ogrisel replied:
> Please, let me know what you think of it.

LGTM!

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+        # end: with nogil, parallel
+        return
+
+    @final
+    cdef void _parallel_on_Y(self) nogil:
```
Do we really need to define this function twice? Or can we parametrize it like
```python
def void _parrallel_on_chunk(is_X_parallel):
```
But if that makes it only more complicated, let's keep the code as is.

jjerphan replied:
I think it helps properly separating concerns. Some reasons I see:
 - the two strategies are different (see the docstring) and having an interfaces for each of them clearly make this explicit
 - callbacks for the two strategies (though sharing pattern of the naming pattern) are different and aren't interchangeable

lorentzenchr replied:
I see. Could you add a comment in the docstring or elsewhere, why there two methods are different, in particular the order of `prange` and `range`.

jjerphan replied:
The difference between both strategies is explained in `PairwiseDistancesReduction`'s docstring:

https://github.com/scikit-learn/scikit-learn/blob/fb927e74bc8ac328cbfde3cbf75f2985ca837569/sklearn/metrics/_pairwise_distances_reduction.pyx#L123-L141

Is this sufficient to you, or should I add more information?

ogrisel replied:
Maybe it would help in the beginning of the paragraph to make it explicit than in both cases with have a loop on Y chunks nested under a loop on X chunks, and that the choice of the strategy will impact which will be made parallel with the help of the Cython/OpenMP `prange` construct.

lorentzenchr commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+    ) nogil:
+        """Compute the pairwise distances on two chunks of X and Y and reduce them.
+
+        This is the core critical region of PairwiseDistanceReductions' computations
```
```suggestion
        This is THE core computation method of PairwiseDistanceReductions'
```

lorentzenchr reviewed (commented):
A pass on `cdef class PairwiseDistancesReduction`.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+    return row_norms
+
+cdef class PairwiseDistancesReduction:
+    """Abstract base class for pairwise distance computation & reduction
```
I accept nitpicks as I've been nitpicking more than you in my reviews for your PRs. :upside_down_face:

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+# Pairwise Distances Reductions
```
I think using the plural for "distance" makes sense here as "Pairwise Distances Reductions" is a compound noun.

jjerphan committed:
DEBUG TST Try removing handling of unstable OpenBLAS configuration

This was introduced once a long time ago for a failure
which was happening in a single configuration (see the comments).

Let's see if this has been fixed.

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+    This is faster than using np.einsum("ij, ij->i") even when using a single thread.
+    """
+    cdef:
+        # Casting for X to remove the const qualifier is needed because APIs
+        # exposed via scipy.linalg.cython_blas aren't reflecting the arguments'
+        # const qualifier.
```
https://github.com/scipy/scipy/issues/14262, that I will add in comments.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,821 @@
+        DTYPE_t[::1] row_norms = np.empty(n, dtype=DTYPE)
+
+    for idx in prange(n, schedule='static', nogil=True, num_threads=num_threads):
+        row_norms[idx] = _dot(d, X_ptr + idx * d, 1, X_ptr + idx * d, 1)
```
I have not: BLAS LEVEL1 operations (used under the hood by `_dot`) have been included to avoid their loopy counterparts. See page 3 of https://dl.acm.org/doi/10.1145/355841.355847 for reference.

lorentzenchr replied:
As info: There is a culture that 👍 in a review comment means that the raised issue has been addressed.

jjerphan committed:
Address review comments

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

jjerphan committed:
TST Remove useless mahalanobis case

jjerphan commented on sklearn/utils/__init__.py:
```diff
@@ -81,6 +81,39 @@
 _IS_32BIT = 8 * struct.calcsize("P") == 32
 
 
+def _in_unstable_openblas_configuration():
```
This is a old fixture. This is removed in 5d7ea09.

jjerphan committed:
Factor the logic for computing last chunks indices

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+            sizeof(ITYPE_t *) * self.chunks_n_threads
+        )
+
+        # Main heaps used by PairwiseDistancesArgKmin._compute to return results.
```
```suggestion
        # Main heaps which will be returned as results by `PairwiseDistancesReduction.compute`.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+            DTYPE_t *heaps_r_distances = self.heaps_r_distances_chunks[thread_num]
+            ITYPE_t *heaps_indices = self.heaps_indices_chunks[thread_num]
+
+        # Pushing the distance and their associated indices on heaps
+        # which keep tracks of the argkmin.
```
```suggestion
        # Pushing the distances and their associated indices on a heap
        # which by construction will keep track of the argkmin.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+        cdef:
+            ITYPE_t idx, jdx
+
+        # Sorting indices of the argkmin for each query vector of X
```
```suggestion
        # Sorting the main heaps portion associated to `X[X_start:X_end]`
        # in ascending order w.r.t the distances.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+          synchronisation.
+
+          - 'auto' relies on a simple heuristic to choose between
+          'parallel_on_X' and 'parallel_on_Y'.
```
Maybe it would also help to give some insights on the performance impacts of this choice:

```suggestion
          'parallel_on_X' and 'parallel_on_Y': when `X.shape[0]` is large enough,
          'parallel_on_X' is usually the most efficient strategy. When `X.shape[0]`
          is small but `Y.shape[0]` is large, 'parallel_on_Y' brings more opportunity
          for parallelism and is therefore more efficient despite the synchronization
          step at each iteration of the outer loop on chunks of `X`.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+          embarrassingly parallel and comes with no datastructures synchronisation.
+
+          - 'parallel_on_Y' dispatches chunks of Y uniformly on threads.
+          Each thread then iterates on all the chunks of X. This strategy is
+          embarrassingly parallel but uses intermediate datastructures
+          synchronisation.
```
```suggestion
          Each thread processes all the chunks of X in turn. This strategy is
          a sequence of embarrassingly parallels subtasks (the inner loop on Y
          chunks) with intermediate datastructures synchronisation at each
          iteration of the sequential outer loop on X chunks.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+            with nogil, parallel(num_threads=self.chunks_n_threads):
+                thread_num = _openmp_thread_num()
+
+                # Initializing datastructures used in this thread
```
```suggestion
                # Initializing thread local datastructures
```

jjerphan replied:
Those aren't thread-local data-structures (because they are accessible outside of some other thread) but are used as if they were.

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+            ITYPE_t Y_start, Y_end, X_start, X_end, X_chunk_idx, Y_chunk_idx
+            ITYPE_t thread_num
+
+        # Allocating datastructures
```
```suggestion
        # Allocating datastructures shared by all threads
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+            ITYPE_t heaps_size = self.X_n_samples_chunk * self.k
+            ITYPE_t thread_num
+
+        # The allocation is done in parallel for data locality purposes: this way
+        # the heaps used in each threads are allocated in pages which are closer
+        # to processor core used by the thread.
```
```suggestion
        # The allocation is done in parallel for data locality purposes: this way
        # the heaps used in each threads are allocated in pages which are closer
        # to the CPU core used by the thread.
        # See comments about First Touch Placement Policy:
        # https://www.openmp.org/wp-content/uploads/openmp-webinar-vanderPas-20210318.pdf #noqa
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+    @final
+    cdef void _parallel_on_X(self) nogil:
+        """Compute the pairwise distances of each row vector of X on Y
+        by parallelizing computation on chunks of X and reduce them.
```
```suggestion
        by parallelizing computation on the outer loop on chunks of X
        and reduce them.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+        with nogil, parallel(num_threads=self.effective_n_threads):
+            # Synchronising the thread heaps with the main heaps.
+            # This is done in parallel sample-wise (no need for locks).
+            # This might break each thread's data locality a bit but
+            # but this is negligible and this parallel pattern has
+            # shown to be efficient in practice.
```
```suggestion
            # 
            # This might break each thread's data locality as each heap which
            # was allocated in a thread is being now being used in several threads.
            # 
            # Still, this parallel pattern has shown to be efficient in practice.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+    @final
+    cdef void _parallel_on_Y(self) nogil:
+        """Compute the pairwise distances of each row vector of X on Y
+        by parallelizing computation on chunks of Y and reduce them.
```
```suggestion
        by parallelizing computation on the inner loop on chunks of Y
        and reduce them.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+
+    strategy : str, {'auto', 'parallel_on_X', 'parallel_on_Y'}, default=None
+        The chunking strategy defining which dataset parallelization are made on.
+
```
```suggestion

        For both strategies the computations happens with two nested loops,
        respectively on chunks of X and chunks of Y.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+    strategy : str, {'auto', 'parallel_on_X', 'parallel_on_Y'}, default=None
+        The chunking strategy defining which dataset parallelization are made on.
+
+        Strategies differs on the dispatching they use for chunks on threads:
```
```suggestion
        Strategies differs on which loop (outer or inner) is made to run
        in parallel with the Cython `prange` construct:
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,829 @@
+                free(self.heaps_r_distances_chunks[thread_num])
+                free(self.heaps_indices_chunks[thread_num])
+
+            # Sort the main heaps into arrays in parallel
+            # in ascending order w.r.t the distances
```
```suggestion
            # Sorting the main in ascending order w.r.t the distances.
            # This is done in parallel sample-wise (no need for locks). 
```

jjerphan reviewed (commented):
Add "I-hope-useful" precision for some comments regarding parallel sections.

ogrisel reviewed (commented):
Here is a batch of suggestions to help with https://github.com/scikit-learn/scikit-learn/pull/22064/files#r774754046.

If the changes suggested to improve the docstring on the `strategy` parameter are accepted, the `_parallel_on_X` and `_parallel_on_Y` docstrings should be updated accordingly.

jjerphan committed:
Improve comments regarding strategies and parallel sections

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,160 @@ cdef class PyFuncDistance(DistanceMetric):
+                **(metric_kwargs or {})
+            )
+
+        if not(X.dtype == Y.dtype == np.float64):
+            raise ValueError("Only 64bit float datasets are supported for X and Y.")
```
Let's be explicit that 32bit can be implemented in the future:

```suggestion
        if not(X.dtype == Y.dtype == np.float64):
            raise ValueError(
                f"Only 64bit float datasets are supported at this time, "
                f"got: X.dtype={X.dtype} and Y.dtype={Y.dtype}"
            )
```

ogrisel commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,160 @@ cdef class PyFuncDistance(DistanceMetric):
+        # Metric-specific checks that do not replace nor duplicate `check_array`.
+        distance_metric._validate_data(X)
+        distance_metric._validate_data(Y)
+
```
```suggestion

    # TODO: dispatch to other dataset pairs for sparse support once
    # available:
```

ogrisel commented on sklearn/metrics/_dist_metrics.pyx:
```diff
@@ -1150,3 +1165,160 @@ cdef class PyFuncDistance(DistanceMetric):
+            If provided as a ndarray, it must be C-contiguous.
+            If provided as a sparse matrix, it must be in CSR format.
+
+        metric : str, default='euclidean'
+            The distance metric to use for argkmin. The default metric is
+            a fast implementation of the standard Euclidean metric.
+            For a list of available metrics, see the documentation of
+            :class:`~sklearn.metrics.DistanceMetric`.
```
```suggestion
        metric : str, default='euclidean'
            The distance metric to compute between rows of X and Y.
            The default metric is a fast implementation of the Euclidean
            metric. For a list of available metrics, see the documentation
            of :class:`~sklearn.metrics.DistanceMetric`.
```

ogrisel reviewed (commented):
Another pass of review. Once addressed (especially the comment on the tests), I think this PR LGTM.

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,844 @@
+                thread_num = _openmp_thread_num()
+
+                # Initializing datastructures used in this thread
+                self._parallel_on_Y_init(thread_num)
```
Naming wise, I think:

```suggestion
                self._parallel_on_Y_parallel_init(thread_num)
```

since this is used in the "parallel" scope.

Which means we rename:

```python
# Allocating datastructures shared by all threads
self._parallel_on_Y_init()
```

which will be cleaned up by `self._parallel_on_Y_finalize()`.

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,844 @@
+        #   - when parallelizing on Y, those heaps pointer heaps are referencing
+        #   small heaps which are thread-wise-allocated and whose content will be
+        #   merged with the main heaps'.
+        self.heaps_r_distances_chunks = <DTYPE_t **> malloc(
+            sizeof(DTYPE_t *) * self.chunks_n_threads
+        )
+        self.heaps_indices_chunks = <ITYPE_t **> malloc(
+            sizeof(ITYPE_t *) * self.chunks_n_threads
+        )
```
Technically, I think these `mallocs` needs to be in `__cinit__`.

jjerphan replied:
That's a good suggestions. To me, this is the kind of pattern that are hard to understand and manage in Cython.

Addressed in 09a95272b4bd5b02692e8fde8ff26959f9814cae

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,844 @@
+np.import_array()
+
+
+cpdef DTYPE_t[::1] _sqeuclidean_row_norms(
+    const DTYPE_t[:, ::1] X,
+    ITYPE_t num_threads,
+):
```
I do not think we make use of `_sqeuclidean_row_norms` in this PR.

thomasjpfan reviewed (approved):
Minor comments, but I think this is ready to go.

There are two paths for follow up:

1. FastEuclideanPairwiseArgKmin
2. Interfacing implementation with scikit-learn interfaces 

I am in favor of option 2 first, because with option 2 completed, we are merged directly into `main`. `FastEuclideanPairwiseArgKmin` would be a follow up performance improvement.

jjerphan committed:
Address reviews' comments

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Remove unused _sqeuclidean_row_norms

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,844 @@
+            # We need to recompute distances because we relied on
+            # surrogate distances for the reduction.
+            self.compute_exact_distances()
+            return np.asarray(self.argkmin_distances), np.asarray(self.argkmin_indices)
```
I believe this is a good point. :+1:

jjerphan committed:
Swap argkmin_indices and argkmin_distances

To have argkmin_indices always be the first
for consistency.

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan commented on sklearn/metrics/tests/test_pairwise_distances_reduction.py:
```diff
@@ -0,0 +1,362 @@
+    )
+
+    dist, indices = PairwiseDistancesReduction.compute(
+        X + 0,
+        Y + 0,
```
Nice catch. I completely forgot about that. :open_mouth: 

> Instead of such a generic test that should work for all reductions, I think we could have dedicated tests for each reduction. For instance, a test that compares the argkmin reduction with a naive scipy cdist + numpy implementation for all supported metrics on data close enough to 0.0 to put ourselves in a favorable numerical setting.

That's a good point. Let's add it in another PR?

jjerphan replied:
Mainly callable functions and specialization of distance metrics present in `PAIRWISE_DISTANCE_FUNCTIONS`:

https://github.com/scikit-learn/scikit-learn/blob/d5e045fd600bf7f3edd984fe7f8599124a67f7f6/sklearn/metrics/pairwise.py#L1372-L1385

ogrisel replied:
Alright, let's focus on the nominal cases for this this test.

jjerphan committed:
Move initializations from __init__ to __cinit__

This is more appropriate, especially for dynamic allocation.

See:
cython.readthedocs.io/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init

    The __cinit__() method is where you should perform basic C-level
    initialisation of the object, including allocation of any C data
    structures that your object will own.

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

ogrisel commented:
> FastEuclideanPairwiseArgKmin would be a follow up performance improvement.

It might actually be needed to avoid a performance regression on machines with few cores as the current numpy code already uses the GEMM trick.

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,834 @@
+
+cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
+    """Compute the argkmin of row vectors of X on the ones of Y.
+
```
```suggestion

    For each row of X, computes the indices of the rows of Y with the smallest
    distances. This is typically used perform bruteforce k-nearest neighbors
    queries.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,834 @@
+
+            Strategies differs on the dispatching they use for chunks on threads:
+
+              - 'parallel_on__X' dispatches chunks of X uniformly on threads.
```
typo

```suggestion
              - 'parallel_on_X' dispatches chunks of X uniformly on threads.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,844 @@
+        by parallelizing computation on the inner loop on chunks of Y
+        and reduce them.
+
+        This strategy dispatches chunks of Y uniformly on threads.
+        Each thread then iterates on all the chunks of X. This strategy is
+        embarrassingly parallel but uses intermediate datastructures
+        synchronisation.
```
It seems that this suggestion was marked resolved without being taken into account or discussed.

jjerphan replied:
Oh yeah, I missed it in my batch. :+1:

ogrisel reviewed (approved):
+1 for merge once the tests have been updated to remove the translation invariance check that is too complicated to get working in general and replaced by a unit test to compare the ArgKMin implementation (with both parallelism strategies) with an naive implementation based on cdist/argpartition on small-ish random data as discussed in  https://github.com/scikit-learn/scikit-learn/pull/22064/files#r777541915

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -0,0 +1,834 @@
+              synchronisation.
+
+              - 'auto' relies on a simple heuristic to choose between
+              'parallel_on__X' and 'parallel_on_Y'.
```
This docstring would benefit from being kept in think with:

https://github.com/scikit-learn/scikit-learn/pull/22064/files#diff-fc1f4e2044627e9eda65f62fbbe018a3821ee87e75ea68c90d64620454f9ebd2R110-R121

Not sure how to handle the redundancy. But I think it's fine to keep the copy.

jjerphan committed:
Improve docstring comment

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Improve comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Add 'pairwise_dist_chunk_size' to scikit-learn config

jjerphan committed:
TST Adapt test for PairwiseDistancesArgKmin translation invariance

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan commented:
> Minor comments, but I think this is ready to go.
> 
> There are two paths for follow up:
> 
>     1. FastEuclideanPairwiseArgKmin
> 
>     2. Interfacing implementation with scikit-learn interfaces
> 
> 
> I am in favor of option 2 first, because with option 2 completed, we are merged directly into `main`. `FastEuclideanPairwiseArgKmin` would be a follow up performance improvement.

So do I. :+1:

ogrisel committed:
test_pairwise_distances_argkmin

ogrisel commented:
@jjerphan here is what I had in mind for the test: https://github.com/jjerphan/scikit-learn/pull/7

ogrisel committed:
Simpler variable names

jjerphan committed:
Merge pull request #7 from ogrisel/test_pairwise_distances_argkmin

jjerphan committed:
fixup! TST Adapt test for PairwiseDistancesArgKmin translation invariance

lorentzenchr reviewed (commented):
A typo from an unfinished further review pass. Thanks @ogrisel and @thomasjpfan for finishing your reviews. I got busy these days.

lorentzenchr commented:
@jjerphan The failing CI due to several `DeprecationWarning` is maybe the missing piece before merge. Looking forward to it!

jjerphan committed:
fixup! fixup! TST Adapt test for PairwiseDistancesArgKmin translation invariance

jjerphan committed:
Use correct orthograph for 'Callback'

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

thomasjpfan commented:
> It might actually be needed to avoid a performance regression on machines with few cores as the current numpy code already uses the GEMM trick.

If there is a performance regression, then I agree that `FastEuclideanPairwiseArgKmin`  should come first.

ogrisel commented:
CI is blue! Merging!
